subt/offseter.py

```diff
@@ -0,0 +1,104 @@
+from threading import Thread
+
+class Offseter:
+
+    def __init__(self, config, bus):
+        bus.register("pose3d")
+        self.thread = Thread(target=self.run)
+        self.thread.name = bus.name
+        self.bus = bus
+
+    def start(self):
+        self.thread.start()
+
+    def join(self, timeout=None):
+        self.thread.join(timeout=timeout)
+
+    def run(self):
+        origin = None
+        xyz = None
+        while None in (origin, xyz):
+            dt, channel, data = self.bus.listen()
+            #print("#", channel, data)
+            if channel == "origin":
+                if len(data) == 8:
+                    robot_name, x, y, z, qa, qb, qc, qd = data
+                    origin = [x, y, z]
+            elif channel == "pose3d":
+                xyz, orientation = data
+            else:
+                raise RuntimeError(f"unknown channel '{channel}'")
+
+        offset = [o - p for o,p in zip(origin, xyz)]
+
+        while True:
+            dt, channel, data = self.bus.listen()
+            #print("-", channel, data)
+            if channel == "origin" and len(data) == 8:
```

Comment:
here you are updating offset whenever origin is received - this is slightly different behavior than in `subt/main.py` when offset is updated only once at the beginning of each run/enter to the tunnel. Also if the robot does not correctly return to the home position, origin may be not available and you will receive ... "origin error" ... OK, that is why `len(data) == 8` I see it now.

Reply:
It is done like this to seamlessly support reentry into the cave while resetting the accumulated error.

Reply:
At the same time it at least allows the X4 to enter the world in urban - the localization does not work in the start area at all there. So far it seems there are some benefits and no immediate down sides.